### ISSUE TITLE
[MaterializedPostgreSQL] PG replication slot size keep raise 

### ISSUE THREAD

**Issue by Zkuns** [ClickHouse/ClickHouse#47979]
I use MaterializedPostgreSQL to sync the data from PG to Clickhouse. 

My database data does not change often. The slot size will keep raise until the database data has changed. After the Database data change, the ClickHouse will sync this change and the PG replication slot size will be down to zero.

The ClickHouse I use is 23.2.4.12.
The Database version is 15.2

**Comment by kssenii** [comment]
please check: during the time when there are no changes on postgresql side, does clickhouse writes logs like "Advanced LSN up to: ..." ?

**Comment by Zkuns** [comment]
No, there no log like this "Advanced LSN up to: ..." when the data not change

**Comment by Zkuns** [comment]
No, there no log like this "Advanced LSN up to: ..." when the data not change

**Comment by kssenii** [comment]
can you please check then what do you see if you do the query `SELECT * FROM pg_logical_slot_get_changes('slot', NULL, NULL);` in PostgreSQL, where "slot" is a slot name from `pg_replication_slots` which ens with suffix "_ch_replication_slot"?

**Comment by kssenii** [comment]
because on the case when there are no changes, as I saw, only empty BEGIN COMMIT messages are coming, but in the recent version it was taken into account, and pg_replication_slot_advance is done in this case to avoid the slot size rising.

**Comment by Zkuns** [comment]
I running this sql get null:
![image](https://user-images.githubusercontent.com/8684193/227963453-cc442e56-9409-42e8-b99a-0b2fb25516bd.png)

Here is my slot detail: 
![image](https://user-images.githubusercontent.com/8684193/227964454-5c0c2f58-ec35-4804-a824-f421633527d9.png)

**Comment by Zkuns** [comment]
After I run this 'pg_logical_slot_get_binary_changes'. The slot size is also down to zero.

**Comment by Zkuns** [comment]
I have two databases. The one often change has this log. But the data stable one do not have this log.

I running this SQL get null here.
`sql
SELECT replay_lsn, pg_wal_lsn_diff(pg_current_wal_lsn(), replay_lsn) FROM pg_stat_replication
`
![image](https://user-images.githubusercontent.com/8684193/228274558-9d5ca26c-0686-4631-8da8-37dc5e438fd2.png)

> Did you check the log level in the server config? this log requires TRACE level, otherwise, it will not be logged
Yes, my ClickHouse log level is TRACE. I can see this log for the often changed database. But not this log for the stable database.

**Comment by jpchev** [comment]
@Zkuns  can you please give some detail? 

How do you get the replication slot size? 

Do you mean the difference between `confirmed_flush_lsn` and `restart_lsn `is `pg_replication_slots` keeps growing?

**Comment by jpchev** [comment]
I cannot reproduce the issue, I use:

postgresql 16.1, 

the following version of ch 

_ClickHouse client version 23.12.2.59 (official build).
Connecting to localhost:9000 as user default.
Connected to ClickHouse server version 23.12.2._

and the following docker compose file
```

version: '3.7'

services:
  db:
    container_name: db
    image: postgres
    restart: always
    environment:
      POSTGRES_DB: my_database
      POSTGRES_USER: my_user
      POSTGRES_PASSWORD: my_password
    command:
      - "postgres"
      - "-c"
      - "wal_level=logical"

  clickhouse:
    image: clickhouse/clickhouse-server
    user: "101:101"
    container_name: clickhouse
    hostname: clickhouse
```

the I run `docker-compose up`
and in postgresql I run

```
psql -d my_database my_user
create table x(a int);
insert into x values(1);
ALTER TABLE x ADD PRIMARY KEY (a);
```
then in ch

```
CREATE DATABASE postgres_db
ENGINE = MaterializedPostgreSQL('db:5432', 'my_database', 'my_user', 'my_password');
use postgres_db;
```
and now I can see the replicated data (1 row) with

`select * from x;`

now I check the replication slots in postgresql and I see there is one created by ch (the slot is named _my_database_)

```
my_database=# \x
Expanded display is on.
my_database=# SELECT * FROM pg_replication_slots;
-[ RECORD 1 ]-------+------------
**slot_name           | my_database**
plugin              | pgoutput
slot_type           | logical
datoid              | 16384
database            | my_database
temporary           | f
active              | f
active_pid          |
xmin                |
catalog_xmin        | 749
**restart_lsn         | 0/1981A90
confirmed_flush_lsn | 0/1981A90**
wal_status          | reserved
safe_wal_size       |
two_phase           | f
conflicting         | f
```

and the following query (the same as in clickhouse code) to see changes in the replication slot without consuming them.

```
my_database=# select * from pg_logical_slot_peek_binary_changes('my_database', NULL, 10, 'publication_names', 'my_database_ch_publication', 'proto_version', '1');
(0 rows)
```

I keep running this latest query over again, yet it returns 0 rows. I tried also to perform empty transactions on postgres, but the query always returns 0 rows.

```
my_database=# begin;
BEGIN
my_database=*# commit;
COMMIT
my_database=# end;
WARNING:  there is no transaction in progress
COMMIT
my_database=# select * from pg_logical_slot_peek_binary_changes('my_database', NULL, 10, 'publication_names', 'my_database_ch_publication', 'proto_version', '1');
(0 rows)
```

**Comment by kssenii** [comment]
> I think this issue might happen in postgresql < 15, as Postgresql 15 has improved the logical replication by avoiding the replication of empty transactions

yeah, this might be the case, though I assumed that this problem with empty transactions should have been fixed by https://github.com/ClickHouse/ClickHouse/pull/40511/commits/b2b97e53ea3e319cea39262aa23eb4378fb8528f#diff-99ffe321e146d9061ab12679d2ef9316b5349dcb87b372d1be3efc888f4722b8R839-R845.


> Do you mean the difference between confirmed_flush_lsn and restart_lsn  in pg_replication_slots keeps growing?

I thought initially that his problem is `pg_wal_lsn_diff` between `confirmed_flush_lsn` and current wal insert lsn. But in my understanding this difference could grow (I think only) because of replication of empty transactions you mentioned, and I considered as I wrote above this issue to be fixed by the commit I linked.

> @kssenii what do you think?
Is this issue still relevant?

Yes, I think let's close this issue then, I also consider this case to be fixed. Will reopen the issue if @Zkuns or someone else comes and says the issue still exists.
